### PR TITLE
8319987: compilation of sealed classes leads to infinite recursion

### PR DESCRIPTION
**Pull request by vicente-romero-oracle**
The compiler can throw a SOE when trying to compile an incorrect sealed classes hierarchy like:
```
    sealed interface Action permits Add {}
    sealed interface MathOp permits Add {}
    sealed static class Add implements MathOp permits Add {}
```
The error is thrown while trying to prove if two classes are disjoint or not. The proposed solution is to keep a set with the pairs of classes analyzed so far and stop as soon as a the current pair of classes is already in the set.

TIA

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8319987](https://bugs.openjdk.org/browse/JDK-8319987): compilation of sealed classes leads to infinite recursion (**Bug** - P4)

### Reviewers
 * [Jan Lahoda](https://openjdk.org/census#jlahoda) (@lahodaj - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16668/head:pull/16668` \
`$ git checkout pull/16668`

Update a local copy of the PR: \
`$ git checkout pull/16668` \
`$ git pull https://git.openjdk.org/jdk.git pull/16668/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16668`

View PR using the GUI difftool: \
`$ git pr show -t 16668`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16668.diff">https://git.openjdk.org/jdk/pull/16668.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16668#issuecomment-1811790572)

### CODE REVIEW

File: src/jdk.compiler/share/classes/com/sun/tools/javac/code/Types.java [hunk 1]
```diff
@@ -1665,37 +1665,49 @@ public boolean isCastable(Type t, Type s, Warner warn) {
...
-            if (ts.isSealed() || ss.isSealed()) {
-                // permitted subtypes have to be disjoint with the other symbol
-                ClassSymbol sealedOne = ts.isSealed() ? ts : ss;
-                ClassSymbol other = sealedOne == ts ? ss : ts;
-                return sealedOne.permitted.stream().allMatch(sym -> areDisjoint((ClassSymbol)sym, other));
+                if (pairsSeen.contains(newPair)) {
+                    return false;
+                } else {
+                    pairsSeen.add(newPair);
+                }
```
Comment (archiecobbs) [inline src/jdk.compiler/share/classes/com/sun/tools/javac/code/Types.java:1684]: Minor simplification possible here:
```javaa
if (!pairsSeen.add(newPair))
    return;
```

Comment (vicente-romero-oracle) [reply src/jdk.compiler/share/classes/com/sun/tools/javac/code/Types.java:1684]: sure I like it thanks!

**Review by lahodaj — APPROVED** [review]
Looks good to me.

**Pull request closed without merging**